ndt7/server/upload/upload.go

```diff
@@ -0,0 +1,221 @@
+// This function exits when
+//     (1) a fatal error occurs or
+//     (2) the maximum elapsed time for the upload test expires.
+// Since this function has access to BBR stats (if available), it's the
+// right place to stop the test early. The rest of the upload code is supposed
+// to stop receiving when this function signals that we're done by closing
+// the channel.
+func measuringLoop(ctx context.Context, request *http.Request,
+	conn *websocket.Conn, dataDir string, dst chan model.Measurement) {
+	logging.Logger.Debug("Starting measurement loop")
+	defer logging.Logger.Debug("Stopping measurement loop")
+	defer close(dst)
+
+	resultfp, err := results.OpenFor(request, conn, dataDir, "upload")
+	if err != nil {
+		return // error printed already
+	}
+
+	defer iox.ErrorLoggingCloser(resultfp).Close()
+
+	sockfp, err := getConnFileAndPossiblyEnableBBR(conn)
+	if err != nil {
+		return // error printed already
+	}
+
+	defer iox.ErrorLoggingCloser(sockfp).Close()
+	t0 := time.Now()
+	ticker := time.NewTicker(spec.MinMeasurementInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case now := <-ticker.C:
+			elapsed := now.Sub(t0)
+			if elapsed > defaultDuration {
+				logging.Logger.Debug("Upload has run for long enough.")
+				return
+			}
+
+			measurement := model.Measurement{
+				Elapsed: elapsed.Seconds(),
+			}
+
+			err = gatherAndSaveTCPInfoAndBBRInfo(&measurement, sockfp, resultfp)
+			if err != nil {
+				return // error printed already
+			}
+
+			dst <- measurement
+		}
+	}
+}
+
+// startMeasuring runs the measurement loop. This runs in a separate goroutine
+// and emits Measurement events on the returned channel.
+func startMeasuring(ctx context.Context, request *http.Request, conn *websocket.Conn, dataDir string) chan model.Measurement {
+	dst := make(chan model.Measurement)
+	go measuringLoop(ctx, request, conn, dataDir, dst)
+	return dst
+}
+
+// ServeHTTP handles the upload subtest.
+func (ul Handler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	// TODO: factor out this stuff as it's the same for both UL/DL
+	logging.Logger.Debug("Upgrading to WebSockets")
+
+	if request.Header.Get("Sec-WebSocket-Protocol") != spec.SecWebSocketProtocol {
+		warnAndClose(writer, "Missing Sec-WebSocket-Protocol in request")
+		return
+	}
+	headers := http.Header{}
+	headers.Add("Sec-WebSocket-Protocol", spec.SecWebSocketProtocol)
+
+	conn, err := ul.Upgrader.Upgrade(writer, request, headers)
+	if err != nil {
+		warnAndClose(writer, "Cannot UPGRADE to WebSocket")
+		return
+	}
+
+	defer conn.Close()
+
+	// Read limit is set to the smallest allowed payload size.
+	conn.SetReadLimit(spec.MinMaxMessageSize)
+
+	// Start measuring loop
+	logging.Logger.Debug("Starting measuring goroutine")
+	ctx, cancel := context.WithCancel(request.Context())
+	measurements := startMeasuring(ctx, request, conn, ul.DataDir)
+
+	// Make sure we clean up resources
+	defer func() {
```

Comment:
We may consider wanting to put here the `conn.Close()` and check the error here.

Reply:
Second thinking, this is not a good pattern. The best pattern is to `defer conn.Close()` as soon as you know it's a valid instance and not `nil`. Saying this because putting the closing of the connection inside of the func is conducive to possible leaks if you don't notice and introduce some new code between the point where the conn is valid and the point in which the func is deferred. This vouches in favour of the approach that I criticised earlier of having some way of knowing if you're ignoring a close. (I still think that, when you know you don't want that to be an error, you probably want an explicitly deferred lambda with proper error checking, however the `iox` interface is a sled between doing nothing and going all in with a full lambda.)

Do you, @evfirerob, or @pboothe have any counter arguments or objections? Otherwise, we can probably resolve this branch of the conversation and move forward.